pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/HLRealtimeSegmentDataManager.java

```diff
@@ -169,9 +169,9 @@ public HLRealtimeSegmentDataManager(final RealtimeSegmentZKMetadata realtimeSegm
     // create and init stream level consumer
     StreamConsumerFactory streamConsumerFactory = StreamConsumerFactoryProvider.create(_streamConfig);
     String clientId = HLRealtimeSegmentDataManager.class.getSimpleName() + "-" + _streamConfig.getTopicName();
-    _streamLevelConsumer = streamConsumerFactory
-        .createStreamLevelConsumer(clientId, _tableNameWithType, SchemaUtils.extractSourceFields(schema),
-            instanceMetadata.getGroupId(_tableNameWithType));
+    _streamLevelConsumer = streamConsumerFactory.createStreamLevelConsumer(clientId, _tableNameWithType,
+        IngestionUtils.getFieldsForRecordExtractor(tableConfig, schema),
```

Comment:
I think I have the answer for my question. I think this is because we may have filter expressions on columns that are not present in the schema.
If this explanation is right, can you please document it some place? (Perhaps in IngestionUtils class?) Also, it may be worthwhile to make two separate calls, and let the caller decide whether to union the fields or otherwise use them independently. Not sure if there is a use case, but think about it. It certainly makes the code more readable. getFieldsToExtractFromStream(), and getFieldsToFilterOn().
Lastly, is it enough to pass in the IngestionConfig (or, maybe even just the FilterConfig) rather than entire TableConfig? This will keep the discipline of adding any ingestion related item to IngestionConfig than looking some place else in the TableConfig

Reply:
1. Yes the explanation is right. Updated javadoc on IngestionUtils#getFieldsForRecordExtractor method.

2. Individual calls should never have to be made, I feel it is better if provide only the method which will extract all the necessary fields, so that we avoid errors in the following phases of ingestion.

3. I like the suggestion of passing just IngestionConfig to set the discipline. Changed the method to take IngestionConfig.